File: ios_storyboard/TodoApp/Features/Todo/TodoListViewModel.swift
import Foundation

// UIKit では @Observable 非対応のため onUpdate クロージャで通知
@MainActor
final class TodoListViewModel {
    private let repository: TodoRepositoryInterface
    var onUpdate: (() -> Void)?

    private(set) var todos: [Todo] = []
    var filter: TodoFilter = .all {
        didSet { onUpdate?() }
    }

    var filteredTodos: [Todo] {
        switch filter {
        case .all: todos
        case .active: todos.filter { !$0.isCompleted }
        case .completed: todos.filter { $0.isCompleted }
        }
    }

    init(repository: TodoRepositoryInterface = TodoRepository()) {
        self.repository = repository
    }

    func loadTodos() {
        todos = repository.fetchAll()
        onUpdate?()
    }

    func toggle(id: String) {
        repository.toggle(id: id)
        loadTodos()
    }

    func delete(id: String) {
        repository.delete(id: id)
        loadTodos()
    }

    func save(_ todo: Todo) {
        repository.save(todo)
        loadTodos()
    }

    func syncFromAPI() {
        Task { @MainActor [weak self] in
            guard let self else { return }
            try? await self.repository.syncFromAPI()
            self.loadTodos()
        }
    }
}

File: android_compose/app/src/main/java/com/example/myapplication/features/todo/TodoListViewModel.kt
package com.example.myapplication.features.todo

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class TodoListState(
    val todos: List<Todo> = emptyList(),
    val filter: TodoFilter = TodoFilter.ALL,
    val isLoading: Boolean = true,
    val isSyncing: Boolean = false,
    val error: String? = null,
) {
    val filtered: List<Todo>
        get() = when (filter) {
            TodoFilter.ALL       -> todos
            TodoFilter.ACTIVE    -> todos.filter { !it.isCompleted }
            TodoFilter.COMPLETED -> todos.filter { it.isCompleted }
        }
}

class TodoListViewModel(
    private val repository: TodoRepositoryInterface,
) : ViewModel() {

    private val _state = MutableStateFlow(TodoListState())
    val state: StateFlow<TodoListState> = _state.asStateFlow()

    init {
        viewModelScope.launch {
            repository.observeAll().collect { todos ->
                _state.update { it.copy(todos = todos, isLoading = false) }
            }
        }
    }

    fun setFilter(filter: TodoFilter) {
        _state.update { it.copy(filter = filter) }
    }

    fun toggle(id: String) {
        viewModelScope.launch {
            val todo = _state.value.todos.find { it.id == id } ?: return@launch
            repository.update(todo.copy(isCompleted = !todo.isCompleted))
        }
    }

    fun delete(id: String) {
        viewModelScope.launch { repository.delete(id) }
    }

    fun sync() {
        viewModelScope.launch {
            _state.update { it.copy(isSyncing = true, error = null) }
            try {
                repository.sync()
            } catch (e: Exception) {
                _state.update { it.copy(error = e.message) }
            } finally {
                _state.update { it.copy(isSyncing = false) }
            }
        }
    }

    class Factory(private val repository: TodoRepositoryInterface) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T =
            TodoListViewModel(repository) as T
    }
}
